Currently, compiler is not using the user defined conversion operator ( i.e. https://github.com/cms-externals/boost/blob/cms/v1.80.0/boost/archive/basic_archive.hpp#L101-L103 ,  https://github.com/cms-externals/boost/blob/cms/v1.80.0/boost/archive/basic_archive.hpp#L139-L141 etc.) and directly cast the object to the type ` (typename boost::uint_t<sizeof(T) * CHAR_BIT>::least&)(t)` [here](https://github.com/cms-sw/cmssw/blob/master/CondFormats/Serialization/interface/eos/portable_iarchive.hpp#L412C6-L412C71)  but generates warning that it did not use the user defined cast operator. This cast works as the classes e.g ` class_id_type, object_id_type` only have one data member of type base_type. We can live with it and ignore the warning using `#pragma GCC diagnostic ignored "-Wcast-user-defined"` ( as proposed by this PR) or we do something like the following
- move `base_type` https://github.com/cms-externals/boost/blob/cms/v1.80.0/boost/archive/basic_archive.hpp#L116  to public section
- use `T::base_type&` which should call the user defined conversion operatorfor class T e.g. for `(typename boost::uint_t<sizeof(T) * CHAR_BIT>::least&)(t)` something like `(typename boost::uint_t<sizeof(T) * CHAR_BIT>::least&)((typename T::b ase_type&)(t))` should work.Compiler will properly call the conversion operator.

I have tested this using the following code 
```
#include <iostream>
class A {
  public:
    typedef int base_type;
  public:
    A(base_type x){t=x;}
    operator base_type(){
      std::cout <<"operator int\n";
      return t;
    }
    explicit operator base_type&(){
      std::cout <<"operator int&\n";
      return t;
    }
  private:
    base_type t;
};

void func(unsigned int& a, unsigned int v){a=v;}
int main()
{
   A x(2);
   std::cout <<(unsigned int&)(x)<<std::endl;
   func((unsigned int &)x,5);
   std::cout <<(unsigned int&)(x)<<std::endl;
   std::cout<<"==========="<<std::endl;
   std::cout <<(typename A::base_type&)(x)<<std::endl;
   func((unsigned int&)((typename A::base_type&)(x)),10);
   std::cout <<(typename A::base_type&)(x)<<std::endl;
   return 0;
}
```

and at compile time I get
```
>> Compiling  src/x/y/bin/test.cpp
src/x/y/bin/test.cpp: In function 'int main()':
src/x/y/bin/test.cpp:23:33: warning: casting 'A' to 'unsigned int&' does not use 'A::operator base_type()' [-Wcast-user-defined]
   23 |    std::cout <<(unsigned int&)(x)<<std::endl;
      |                                 ^
src/x/y/bin/test.cpp:24:25: warning: casting 'A' to 'unsigned int&' does not use 'A::operator base_type()' [-Wcast-user-defined]
   24 |    func((unsigned int &)x,5);
      |                         ^
src/x/y/bin/test.cpp:25:33: warning: casting 'A' to 'unsigned int&' does not use 'A::operator base_type()' [-Wcast-user-defined]
   25 |    std::cout <<(unsigned int&)(x)<<std::endl;
      |                                 ^
```

and at runtime I get
```
2
5
===========
operator int&
5
operator int&
operator int&
10
```

So either we patch boost and move base_type for these classes to public or we let compiler cast the objectly directly and ignore warning using `#pragma GCC diagnostic...` .

@makortel @Dr15Jones  what do you think?

--- a/CondFormats/Serialization/interface/eos/portable_iarchive.hpp
+++ b/CondFormats/Serialization/interface/eos/portable_iarchive.hpp
@@ -409,7 +409,7 @@ namespace eos {
       // created through BOOST_STRONG_TYPEDEF(X, some unsigned int) like
       // library_version_type, collection_size_type, item_version_type,
       // class_id_type, object_id_type, version_type and tracking_type
-      load((typename boost::uint_t<sizeof(T) * CHAR_BIT>::least&)(t));
+      load(static_cast<T::base_type&>(t));
     }
   };
 

